runtime-modules/proposals/engine/src/lib.rs

```diff
@@ -433,7 +421,7 @@ decl_module! {
 
             // Check vote results. Approved proposals with zero grace period will be
             // transitioned to the PendingExecution status.
-            for  proposal_data in finalized_proposals {
+            for proposal_data in finalized_proposals {
```

Comment:
I don't understand what is happening here, why are you writing the proposal write back, and then running `finalize_proposal`, which mutates it. Moreover`finalize_proposal` does checks like `is_active_or_pending_execution` which trivially must hold given that there is a valid `decision_status` provided.
Later you even go in `get_approved_proposal_ready_for_execution` and reread same proposals from storage.

In general, reading the entire code path in `on_finalize` is not very easy, there are so mange branches with side-effects and stages where things are happening here.

The entire process should be a simple single pure computation `{p | proposal p} => {what to do about proposal p + require static information | proposal p}` and then one big `match` side-effect computation where you take the output of this and do what you need case by case, whether its overwriting an old proposal or deleting one, or executing. Right now I cant easily convince myself that this works, and you are almost certainly avoiding lots of checks and rechecks

Reply:
'Write back a proposal before the finalization' was a harmless bug (I removed it). It is an artifact from the 'archival proposal world'.

The process in the on_finalize() is very similar to your description with one addition: on_finalize also gathers and executes the proposals with an expired grace period. I renamed types and methods and add more comments to improve code readability

Currently, on_finalize contains 4 steps:
1. Get proposals that ready for finalization.
2. Finalize them with side-effects.
3. Get proposals ready for execution (including proposals from the previous step).
4. Execute proposals that ready for execution.

I hope those stages became clear after the last refactoring.